### PR TITLE
fix(platform): add needUpdate func in cluster/machine provider

### LINKED ISSUE #1484: allow user add needUpdate func in cluster/machine provider
企业微信截图_cb5fb9a5-a67d-4336-a260-3f7af565cfda

allow add provider needupdate func in this place.

### PR DESCRIPTION
**Pull request by TomatoAres**
**What type of PR is this?**

> /kind feature

**What this PR does / why we need it**:

User can define  customized `needUpdate` function in provider, default reture false.

**Which issue(s) this PR fixes**:

Fixes #1484

### CODE REVIEW

File: pkg/platform/provider/cluster/interface.go [hunk 1]
```diff
@@ -90,6 +90,9 @@ type CredentialProvider interface {
 type Provider interface {
 	Name() string
 
+	// NeedUpdate could be implemented by user to judge whether cluster need update or not.
+	NeedUpdate(old, new *platformv1.Cluster) bool
```
Comment (leoryu) [inline pkg/platform/provider/cluster/interface.go:94]: move it to ContreollerProvider

File: pkg/platform/provider/cluster/interface.go [hunk 2]
```diff
@@ -90,6 +90,9 @@ type CredentialProvider interface {
...
+
 	APIProvider
 	ControllerProvider
```
Comment (leoryu) [inline pkg/platform/provider/cluster/interface.go:97]: add NeedUpdate Interface ini ti

File: pkg/platform/controller/cluster/cluster_controller.go [hunk 3]
```diff
@@ -131,9 +132,18 @@ func (c *Controller) addCluster(obj interface{}) {
...
+	if err != nil || provider == nil {
+		c.log.Error(err, "checking cluster needUpdate", "provider ", provider)
+		if !c.needsUpdate(oldCluster, cluster) {
+			return
+		}
+	} else {
+		if !c.needsUpdate(oldCluster, cluster) && !provider.NeedUpdate(oldCluster, cluster) {
+			return
+		}
 	}
```
Comment (leoryu) [inline pkg/platform/controller/cluster/cluster_controller.go:145]: controllerNeedUpddateResult := c.needsUpdate(oldCluster, cluster)
var providerNeedUpddateResult bool

provider, _ := clusterprovider.GetProvider(cluster.Spec.Type)
if provider == nil {
  c.log.Infof("get provider with %s failed",cluster.Spec.Type)
} else {
  providerNeedUpddateResult = provider.NeedUpdate(oldCluster, cluster)
}
if !(controllerNeedUpddateResult || providerNeedUpddateResult) {
		return
}

File: pkg/platform/controller/cluster/cluster_controller.go [hunk 4]
```diff
@@ -131,7 +132,17 @@ func (c *Controller) addCluster(obj interface{}) {
...
 	oldCluster := old.(*platformv1.Cluster)
 	cluster := obj.(*platformv1.Cluster)
-	if !c.needsUpdate(oldCluster, cluster) {
+
+	controllerNeedUpddateResult := c.needsUpdate(oldCluster, cluster)
+	var providerNeedUpddateResult bool
+
+	provider, _ := clusterprovider.GetProvider(cluster.Spec.Type)
+	if provider == nil {
+		c.log.Infof("get cluster provider with %s failed", cluster.Spec.Type)
```
Comment (pavlelee) [inline pkg/platform/controller/cluster/cluster_controller.go:141]: modify log to error level. see https://github.com/tkestack/tke/pull/1503/files#diff-303134097085051ccecc6547b55796f466ec830f43c56fcb19736bd3ee2bb35fL304

File: pkg/platform/controller/machine/machine_controller.go [hunk 5]
```diff
@@ -109,7 +109,17 @@ func (c *Controller) addMachine(obj interface{}) {
...
 	oldMachine := old.(*platformv1.Machine)
 	machine := obj.(*platformv1.Machine)
-	if !c.needsUpdate(oldMachine, machine) {
+
+	controllerNeedUpddateResult := c.needsUpdate(oldMachine, machine)
+	var providerNeedUpddateResult bool
+
+	provider, _ := machineprovider.GetProvider(machine.Spec.Type)
+	if provider == nil {
+		c.log.Infof("get machine provider with %s failed", machine.Spec.Type)
```
Comment (pavlelee) [inline pkg/platform/controller/machine/machine_controller.go:118]: remove this log

**Pull request merged**